pkg/lightning/backend/remote/chunk_sender.go

```diff
@@ -0,0 +1,410 @@
+	Finished       bool   `json:"finished"`
+	Error          string `json:"error"`
+}
+
+// chunk is a unit of data that send to remote worker.
+type chunk struct {
+	id   uint64
+	data []byte
+}
+
+type chunksState struct {
+	FlushedChunkID uint64 `json:"flushed-chunk-id"`
+	HandledChunkID uint64 `json:"handled-chunk-id"`
+}
+
+type chunkTask struct {
+	// For flush, resp is used to notify the completion of the flush.
+	// For put chunk, resp is used to receive chunkID allocated by chunkSenderLoop.
+	resp  chan uint64
+	flush bool
+	data  []byte
+}
+
+func newChunkTask(data []byte, flush bool) *chunkTask {
+	return &chunkTask{
+		resp:  make(chan uint64, 1),
+		flush: flush,
+		data:  data,
+	}
+}
+
+func (t *chunkTask) done(chunkID uint64) {
+	t.resp <- chunkID
+	close(t.resp)
+}
+
+type chunkSender struct {
+	id uint64
+	wg sync.WaitGroup
+
+	e           *engine
+	httpClient  *http.Client
+	chunksCache *chunksCache
+	logger      log.Logger
+
+	// state is used to record the state of the chunks in remote worker.
+	state       atomic.Pointer[chunksState]
+	nextChunkID atomic.Uint64
+
+	// the following channels should be closed by chunkSenderLoop
+	loopDoneChan chan struct{} // notify chunkSenderLoop is done
+
+	// the following channels should be closed by closed method
+	taskChan chan *chunkTask
+	cancel   context.CancelFunc
+
+	err error
+}
+
+func newChunkSender(ctx context.Context, id uint64, engine *engine, chunksCache *chunksCache) *chunkSender {
+	// The ctx is used to control the life of the `chunkSenderLoop`.
+	ctx, cancel := context.WithCancel(ctx)
+	c := &chunkSender{
+		id: id,
+		wg: sync.WaitGroup{},
+
+		e:           engine,
+		httpClient:  engine.httpClient,
+		chunksCache: chunksCache,
+		logger:      engine.logger.With(zap.Uint64("sender", id)),
+
+		state:       atomic.Pointer[chunksState]{},
+		nextChunkID: atomic.Uint64{},
+
+		loopDoneChan: make(chan struct{}),
+		taskChan:     make(chan *chunkTask),
+		cancel:       cancel,
+	}
+	c.state.Store(&chunksState{})
+
+	c.wg.Add(1)
+	go c.chunkSenderLoop(ctx)
+
+	c.logger.Info("chunk sender started")
+	return c
+}
+
+func (c *chunkSender) putChunk(ctx context.Context, data []byte) (uint64, error) {
+	task := newChunkTask(slices.Clone(data), false)
+
+	select {
+	case <-ctx.Done():
+		return 0, ctx.Err()
+	case c.taskChan <- task:
+	case <-c.loopDoneChan:
+		return 0, c.err
+	}
+
+	select {
+	case <-ctx.Done():
+		return 0, ctx.Err()
+	case newChunkID := <-task.resp:
+		return newChunkID, nil
+	case <-c.loopDoneChan:
+		return 0, c.err
+	}
+}
+
+func (c *chunkSender) flush(ctx context.Context) error {
+	if c.getFlushedChunkID() == c.getLastChunkID() {
+		return nil
+	}
+
+	task := newChunkTask(nil, true)
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case c.taskChan <- task:
+	case <-c.loopDoneChan:
+		return c.err
+	}
+
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-task.resp:
+	case <-c.loopDoneChan:
+		return c.err
+	}
```

Comment:
this part is similar to line 136~151. You can extract a helper function and the caller of helper function discard `newChunkID` when it's a flush task